File: typescript/packages/core/src/schemas.ts
import { z } from "zod";

const decimalString = z.string().regex(/^\d+(?:\.\d+)?$/);

export const modelPricingSchema = z
  .object({
    inputUsdPerMillion: decimalString,
    cachedInputUsdPerMillion: decimalString,
    outputUsdPerMillion: decimalString,
    providerMarkupBps: z.number().int().min(0).max(100_000).default(0),
    fixedFeeUsd: decimalString.default("0"),
  })
  .strict();

export const pricingTableSchema = z
  .object({
    version: z.string().min(1).max(128),
    models: z.record(z.string().min(1), modelPricingSchema),
  })
  .strict()
  .refine((table) => Object.keys(table.models).length > 0, {
    message: "At least one model must be priced",
  });

export const usageSchema = z
  .object({
    input_tokens: z.number().int().nonnegative(),
    output_tokens: z.number().int().nonnegative(),
    total_tokens: z.number().int().nonnegative().optional(),
    input_tokens_details: z
      .object({ cached_tokens: z.number().int().nonnegative().optional() })
      .passthrough()
      .optional(),
  })
  .passthrough();

export const receiptSchema = z
  .object({
    id: z.string().min(1),
    createdAt: z.string().datetime(),
    network: z.string().min(1),
    asset: z.string().min(1),
    amountAtomic: z.string().regex(/^\d+$/),
    amountUsd: decimalString,
    model: z.string().min(1),
    pricingVersion: z.string().min(1),
    inputTokens: z.number().int().nonnegative(),
    cachedInputTokens: z.number().int().nonnegative(),
    outputTokens: z.number().int().nonnegative(),
    openaiRequestId: z.string().optional(),
  })
  .strict();

export type ModelPricing = z.infer<typeof modelPricingSchema>;
export type PricingTable = z.infer<typeof pricingTableSchema>;
export type TokenUsage = z.infer<typeof usageSchema>;
export type PaymentReceipt = z.infer<typeof receiptSchema>;

File: python/src/x402_openai/core/schemas.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class ModelPricing(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    input_usd_per_million: str = Field(alias="inputUsdPerMillion")
    cached_input_usd_per_million: str = Field(alias="cachedInputUsdPerMillion")
    output_usd_per_million: str = Field(alias="outputUsdPerMillion")
    provider_markup_bps: int = Field(default=0, alias="providerMarkupBps", ge=0, le=100_000)
    fixed_fee_usd: str = Field(default="0", alias="fixedFeeUsd")

    @field_validator(
        "input_usd_per_million",
        "cached_input_usd_per_million",
        "output_usd_per_million",
        "fixed_fee_usd",
    )
    @classmethod
    def validate_decimal(cls, value: str) -> str:
        import re

        if not re.fullmatch(r"\d+(?:\.\d+)?", value):
            raise ValueError("must be a non-negative decimal string")
        return value


class PricingTable(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    version: str = Field(min_length=1, max_length=128)
    models: dict[str, ModelPricing]

    @field_validator("models")
    @classmethod
    def require_models(cls, value: dict[str, ModelPricing]) -> dict[str, ModelPricing]:
        if not value:
            raise ValueError("at least one model must be priced")
        return value


class TokenUsage(BaseModel):
    model_config = ConfigDict(extra="allow", frozen=True)

    input_tokens: int = Field(ge=0)
    output_tokens: int = Field(ge=0)
    cached_input_tokens: int = Field(default=0, ge=0)


class PaymentReceipt(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    created_at: datetime = Field(alias="createdAt")
    network: str
    asset: str
    amount_atomic: str = Field(alias="amountAtomic", pattern=r"^\d+$")
    amount_usd: str = Field(alias="amountUsd")
    model: str
    pricing_version: str = Field(alias="pricingVersion")
    input_tokens: int = Field(alias="inputTokens", ge=0)
    cached_input_tokens: int = Field(alias="cachedInputTokens", ge=0)
    output_tokens: int = Field(alias="outputTokens", ge=0)
    openai_request_id: str | None = Field(default=None, alias="openaiRequestId")
